src/ds/thread_messaging.h

```diff
@@ -0,0 +1,210 @@
+    union
+    {
+      Payload data;
+      uint64_t padding[14];
+    };
+
+    static void check_invariants()
+    {
+      static_assert(
+        sizeof(ThreadMsg) == sizeof(Tmsg<Payload>), "message is too large");
+      static_assert(
+        sizeof(Payload) <= sizeof(ThreadMsg::padding),
+        "message payload is too large");
+
+      static_assert(
+        offsetof(Tmsg, cb) == offsetof(ThreadMsg, cb),
+        "Expected cb at start of struct");
+      static_assert(
+        offsetof(Tmsg, next) == offsetof(ThreadMsg, next),
+        "Expected next after cb in struct");
+      static_assert(
+        offsetof(Tmsg, data) == offsetof(ThreadMsg, padding),
+        "Expected payload after next in struct");
+    }
+  };
+
+  static void init_cb(std::unique_ptr<ThreadMsg> stuff)
+  {
+    LOG_INFO << "Init was called" << std::endl;
+  }
+
+  class Task
+  {
+#ifdef USE_MPSCQ
+    queue::MPSCQ<ThreadMsg> queue;
+#else
+    std::atomic<ThreadMsg*> item_head = nullptr;
+    ThreadMsg* local_msg = nullptr;
+#endif
+
+  public:
+    Task()
+    {
+#ifdef USE_MPSCQ
+      auto msg = new ThreadMsg;
+      msg->cb = &init_cb;
+      queue.init(msg);
+#endif
+    }
+
+    bool run_next_task()
+    {
+#ifdef USE_MPSCQ
+      if (queue.is_empty())
+      {
+        return false;
+      }
+
+      ThreadMsg* current;
+      bool result;
+      std::tie(current, result) = queue.dequeue();
+
+      if (result)
+      {
+        current->cb(std::unique_ptr<ThreadMsg>(current));
+      }
+#else
+      if (local_msg == nullptr && item_head != nullptr)
+      {
+        local_msg = item_head.exchange(nullptr);
+        reverse_local_messages();
+      }
+
+      if (local_msg == nullptr)
+      {
+        return false;
+      }
+
+      ThreadMsg* current = local_msg;
+      local_msg = local_msg->next;
+
+      current->cb(std::unique_ptr<ThreadMsg>(current));
+#endif
+      return true;
+    }
+
+    void add_task(ThreadMsg* item)
+    {
+#ifdef USE_MPSCQ
+      queue.enqueue(item, item);
+#else
+      ThreadMsg* tmp_head;
+      do
+      {
+        tmp_head = item_head.load();
+        item->next = tmp_head;
+      } while (!item_head.compare_exchange_strong(tmp_head, item));
+#endif
+    }
+
+  private:
+#ifndef USE_MPSCQ
+    void reverse_local_messages()
+    {
+      if (local_msg == NULL)
+        return;
+
+      ThreadMsg *prev = NULL, *current = NULL, *next = NULL;
+      current = local_msg;
+      while (current != NULL)
+      {
+        next = current->next;
+        current->next = prev;
+        prev = current;
+        current = next;
+      }
+      // now let the head point at the last node (prev)
+      local_msg = prev;
+    }
+#endif
+  };
+
+  class ThreadMessaging
+  {
+    std::atomic<bool> finished;
+    std::vector<Task> tasks;
+
+  public:
+    static ThreadMessaging thread_messaging;
+    static std::atomic<uint16_t> thread_count;
+
+  public:
+    ThreadMessaging(uint16_t num_threads = 64) :
```

Comment:
We could add a check here to make sure that the number of threads is smaller than `64`.

Reply:
The most important constraint is that it has to be less than the number of TCS configured, and we need a (public) API in OpenEnclave to expose that.

Reply:
as discussed, i will open a ticket in the open enclave repo